### ISSUE TITLE
Allow JS code to hook into system.run()

### ISSUE THREAD

**Issue by fatcerberus** [spheredev/neosphere#76]
Currently when using the `threads` module, a game must call `threads.join()` to enter the event loop.  Calling `system.run()` will poll input devices and settle promises but won't update any threads.

Implement a feature to allow games to define update and render handlers that get called automatically whenever the game calls `system.run()` and `screen.flip()`, respectively.

- [x] `Dispatch.now()` and `Dispatch.later()` queue one-shot jobs, return a JobToken
- [x] `Dispatch.onX()` sets up a recurring job, returns a JobToken
- [x] JobToken allows you to cancel a pending or recurring job

**Comment by fatcerberus** [comment]
There are a few challenges with this.
- There is no rate limiting for deferred actions (`system.dispatch()`).  Because the framelimiter is quite tightly linked with the event pump (we need to pump events while waiting for the next frame), this means that if you just dispatch updates normally, your game ends up updating maybe thousands of times per second, at a variable and unpredictable rate.  No good.
- Rendering.  It's not enough just to `system.dispatch()` renders, because ideally they should be skipped if the frame is skipped.  Even so the render queue must be cleared, since it'll be outdated next frame.  Alternatively, it might be possible just to avoid dispatching renders at all on a skipped frame.
- The Sphere 1.x map engine presents a challenge.  Its API is not suited to asynchronous operation at all so getting it to cooperate with what basically amounts to a coroutine engine will probably not be easy.  While it's an option to just leave it alone and document the limitation, that would be a major barrier to migration, something I want to make as painless as possible.

**Comment by fatcerberus** [comment]
So the interface I came up with revolves around two functions: `system.defer()` and `system.recur()`.  
- `system.defer(hint, fn)` fires only once and is useful in, say, promise implementations where you need the system to call your function later on your behalf.  It's probably not useful to pass anything other than `Defer.ASAP` to this function so the hint argument might just get removed before minisphere 4.3 is released, or maybe changed to be a time delay in seconds (like setTimeout() in the browser).
- `system.recur(hint, fn)` is like Sphere 1.x SetUpdateScript() and SetRenderScript() on steroids.  It sets up a recurring call which is useful for registering render hooks and such.  system.recur() is irreversible in the current implementation, but that won't be true for long.  I want to implement some sort of AsyncToken object that allows you to cancel async jobs.  system.recur() would give you a token, which you could then go `tok.cancel()` on to cancel the job.

**Comment by fatcerberus** [comment]
Alright, the latest build returns a JobToken from `system.defer()` and friends so that you can do:

``` js
var updateJob = system.recur(Dispatch.Update, updateFunc);
// then later on...
updateJob.cancel();
```

which is useful for encapsulation.  One downside is that there's no job priority - jobs are executed in FIFO order always.  Not sure it's worthwhile to add that feature to the core - one can always layer a threader like miniRT's on top if they need it.

**Comment by fatcerberus** [comment]
Thinking about this some more, with async quickly becoming a core feature, the current API seems like it might be kind of awkward.  Some namespacing might be in order rather than shoving everything into two methods of `system` (effectively a catch-all).  I'll give the design some more thought as minisphere 4.3 gets closer to release.

**Comment by fatcerberus** [comment]
Regarding that redesign: The current setup in master namespaces async functionality into the "Dispatch API":
- `Dispatch.now()` runs as soon as possible (on the next tick).
- `Dispatch.later()` is time-delayed like browser `setTimeout()`.
- `Dispatch.onFlip()` runs at the end of a frame, before the screen is flipped.
- `Dispatch.onUpdate()` runs at the start of a frame to update game state.

Each returns a JobToken object which can be used to cancel the call:

``` js
var delayed = Dispatch.later(5.0, function() {
    console.log("5 seconds ago someone thought this was a good idea.");
});

// ...but then it turns out it wasn't really a good idea after all.
delayed.cancel();
```

For all intents and purposes the issue in this thread is now solved.
